crates/bevy_ui/src/render/mod.rs

```diff
@@ -642,7 +655,41 @@ pub fn prepare_uinodes(
         image.id() != DEFAULT_IMAGE_HANDLE.id()
     }
 
-    for extracted_uinode in extracted_uinodes.uinodes.drain(..) {
+    let mut drains: Vec<_> = extracted_uinodes
+        .uinodes
+        .iter_mut()
+        .map(|uinodes| {
+            let mut d = uinodes.drain(..);
+            let first = d.next();
+            (first, d)
+        })
+        .collect();
+
+    let mut next_extracted_node = move || {
+        let mut min_stack_index = usize::MAX;
+        let mut n = usize::MAX;
+
+        for (i, (node, _)) in drains.iter_mut().enumerate() {
+            if let Some(node) = node {
+                if node.stack_index < min_stack_index {
+                    n = i;
+                    min_stack_index = node.stack_index;
+                }
+            }
+        }
+
+        if n == usize::MAX {
+            None
+        } else {
+            let (current, drain) = &mut drains[n];
+            let next = drain.next();
+            let out = current.take();
+            *current = next;
+            out
+        }
+    };
+
+    while let Some(extracted_uinode) = next_extracted_node() {
```

Comment:
As you mentioned in the PR description, this needs to be a method on `ExtractedUiNode`. Ideally an `impl Iterator` `next` method.